[16.0] shopinvader base url: add schemas

qgroulard opened the pull request:
* [x] includes #1390

sebastienbeau committed:
shopinvader_base_url: pre-commit stuff, remove deprecated file and rename module

sebastienbeau committed:
shopinvader_base_url: WIP refactor module

sebastienbeau committed:
shopinvader_base_url: WIP refactor module, test are green

sebastienbeau committed:
shopinvader_base_url: finish refactor, only check url if something have changed

sebastienbeau committed:
shopinvader_product_url: add module, shopinvader_base_url: improve module

qgroulard committed:
shopinvader_base_url: add fields url_key and redirect_url_key, shopinvader_product_url add fields to schemas

lmignon commented:
@qgroulard Can you take a look at the error?

lmignon commented on shopinvader_product_url/schemas/product.py:
```diff
@@ -0,0 +1,18 @@
+    @classmethod
+    def from_product_product(cls, odoo_rec):
+        obj = super().from_product_product(odoo_rec)
+        obj.url_key = odoo_rec.url_key or None
```
@qgroulard IMO we must call the _update_url_key method to ensure that the urls are computed of refreshed

lmignon commented on shopinvader_product_url/models/product_category.py:
```diff
@@ -0,0 +1,33 @@
+            parent_url = self.parent_id._get_main_url(referential, lang)
+            if parent_url:
+                return "/".join([parent_url.key, url_key])
+        return url_key
```
@thibaultrey  @qgroulard @sebastienbeau IMO for shopinvader we must prefix the url by a 'c/' for categories, 'p/' for products .... This is required to allows the website to known which page must be displayed depending of the prefix....

sebastienbeau replied:
Hi, I check with @thibaultrey and it's not mandatory.

When you go on the website, a query in elastic is done searching in both index (categ and product). And depending on the result it return the right template.

Adding a prefix may improve the performance, it's will be interesting to check on huge catalog, if their is significant performance improvement (maybe the impact is totally insignificant)

Note: In some case you do not want to have prefix
- case where you migrate a site and keep the same url (without redirection)
- case where the customer do not want prefix (depend on the SEO consultant)


So I am more in favor of not having a mandatory prefix, and make it optional.

lmignon replied:
@sebastienbeau @thibaultrey  for alcyon we've a prefix to improve perfs and lower loads on SE. We'll keep it simple and don't set a prefix. For  my project I'll override the `_generate_url_key`method to add a prefix.

lmignon replied:
@thibaultrey For alcyon the pfx is useful to manage the routing and knwo wich component must be instantiate according to the url into the router.

lmignon committed:
[FIX] shopinvader_base_url: Add missing dependency

The dependency on base_sparse_field_list_support is required since we use a Serialized field to store list of values

lmignon committed:
[FIX] shopinvader_base_url: Import Serialized from its module

This is required to avoid race condition when running tests if the base_sparse_field addon has not yet patch the odoo.fields namespace to register the Serialized field class

lmignon committed:
[FIX] shopinvader_product_url: Ensure url_key is filled

When serializing a product or a category, ensures the url_key is properly initialized

lmignon committed:
[IMP] shopinvader_product_url: Cleanup manifest

lmignon commented:
/ocabot merge  nobump
